apps/emqx_gcp_device/src/emqx_gcp_device_api.erl

```diff
@@ -0,0 +1,459 @@
+    end.
+
+-spec remove_device(deviceid()) -> {204}.
+remove_device(DeviceId) ->
+    ok = emqx_gcp_device:remove_device(DeviceId),
+    {204}.
+
+%%------------------------------------------------------------------------------
+%% Internal functions
+%%------------------------------------------------------------------------------
+
+-define(KEY_TYPES, [<<"RSA_PEM">>, <<"RSA_X509_PEM">>, <<"ES256_PEM">>, <<"ES256_X509_PEM">>]).
+
+-spec deviceid() -> tuple().
+deviceid() ->
+    deviceid(#{}).
+
+-spec deviceid(map()) -> tuple().
+deviceid(Override) ->
+    {deviceid,
+        hoconsc:mk(
+            binary(),
+            maps:merge(
+                #{
+                    desc => ?DESC(deviceid),
+                    required => true,
+                    example => <<"c2-ec-x509">>
+                },
+                Override
+            )
+        )}.
+
+-spec validate_devices([imported_device()]) -> {ok, [validated_device()]} | {error, binary()}.
+validate_devices(Devices) ->
+    validate_devices(Devices, []).
+
+-spec validate_devices([imported_device()], [validated_device()]) ->
+    {ok, [validated_device()]} | {error, binary()}.
+validate_devices([], Validated) ->
+    {ok, lists:reverse(Validated)};
+validate_devices([Device | Devices], Validated) ->
+    case validate_device(Device) of
+        {ok, ValidatedDevice} ->
+            validate_devices(Devices, [ValidatedDevice | Validated]);
+        {error, _} = Error ->
+            Error
+    end.
+
+-spec validate_device(imported_device()) -> {ok, validated_device()} | {error, binary()}.
+validate_device(Device) ->
+    validate([deviceid, project, location, registry, keys, config], Device).
+
+-spec validate([checked_device_fields()], imported_device()) ->
+    {ok, validated_device()} | {error, binary()}.
+validate(Fields, Device) ->
+    validate(Fields, Device, #{}).
+
+-spec validate([checked_device_fields()], imported_device(), validated_device()) ->
+    {ok, validated_device()} | {error, binary()}.
+validate([], _Device, Validated) ->
+    {ok, Validated};
+validate([key_type | Fields], #{<<"key_type">> := KeyType} = Device, Validated) ->
+    case lists:member(KeyType, ?KEY_TYPES) of
+        true ->
+            validate(Fields, Device, Validated#{key_type => KeyType});
+        false ->
+            {error, <<"invalid key_type">>}
+    end;
+validate([key | Fields], #{<<"key">> := Key} = Device, Validated) ->
+    validate(Fields, Device, Validated#{key => Key});
+validate([expires_at | Fields], #{<<"expires_at">> := Expire} = Device, Validated) when
+    is_integer(Expire)
+->
+    validate(Fields, Device, Validated#{expires_at => Expire});
+validate([expires_at | _Fields], #{<<"expires_at">> := _}, _Validated) ->
+    {error, <<"invalid expires_at">>};
+validate([expires_at | Fields], Device, Validated) ->
+    validate(Fields, Device, Validated#{expires_at => 0});
+validate([Field | Fields], Device, Validated) when Field =:= deviceid; Field =:= key ->
```

Comment:
`Field =:= key` has matched in the above clause (L401)

Reply:
But at L401 it also checks for <<"key">> being present. If this field is not present, it'll match this case on L411.